### PR TITLE
[1055]: recurring invoices automated date insertion of relative date definitions

### LINKED ISSUE #1055: [Feature Request]: Recurring invoices - automated date insertion of relative date definitions
Recurring invoices - automated date insertion of a relative date definition

## Expected Behavior

Recurring invoices shall have a replacement on date templates/relative date definitions which are set in either the product itself or in the based invoice. Following examples are for the German area (but should be converted to others easily):
Examples:
* an invoice item text part {{{month}}} should be replaced with the actual month: Jun.
* an invoice item text part {{{year+1}}} should be replaced with the incremented year: 2025
* etc.

## Current Behavior

Static text, and not very flexible for recurring invoices

## Possible Solution

Allow tags within products {{{month+1}}} {{{year+1}}} etc. That is possible anyway right now.
Before creating the static PDF, replace this text with a useful meaning, like

> increment a date
* From: Invoiced period {{{month+1}}} - {{{month+3}}} {{{year}}}
* To: Invoiced period Jul. - Aug. 2024
> decrement a date
* From: Past period {{{month-2}}} - {{{month}}} {{{year}}}
* To: Past period Apr. - Jun. 2024

> watch the year overflow at month+3 in December which have to increase year too.

## Context (Environment)

I have a lot of invoices where I have to put into the item a text stating the accounting period and I have some yearly recurring invoices.

## Detailed Description

The change would have to take place direct before the static PDF is created and send to a customer, converting the replacement part to static text in the clients/customers language. Indeed this functionality should not be restricted to recurring invoices.

## Possible Implementation

Right now I have it this way in views/invoice_templates/pdf/my_template.php within 'table class="item-table"' and this way the pdf is in it's final shape before sending

### PR DESCRIPTION
**Pull request by marneu**
## Related Issue
[#1055](https://github.com/InvoicePlane/InvoicePlane/issues/1055)

## Motivation and Context
Within recurring invoices one may need relative date definitions on the PDF per item.
Examples: {{{Date}}} {Month-3}}}-{{{Month+1}}} {{{Year}}}

## Pull Request Checklist

  * [?] My code follows the code formatting guidelines.
  * [X] I have an issue ID for this pull request.
  * [X] I selected the corresponding branch.
  * [X] I have rebased my changes on top of the corresponding branch.

## Issue Type (Please check one or more)

  * [ ] Bugfix
  * [x] Improvement of an existing Feature
  * [x] New Feature

### CODE REVIEW

File: application/helpers/invoice_helper.php [hunk 1]
```diff
@@ -128,4 +129,113 @@ function invoice_qrcode($invoice_id) {
...
+
+            // replace within item description
+            $item_description = str_replace($replaceThis, $withReplacement, $item_description);
+
+        } catch (Exception $e) {
+            echo 'ERROR Message: ' .$e->getMessage();
+        }
+                
+    return $item_description;
+}
```
Comment (nielsdrost7) [inline application/helpers/invoice_helper.php:241]: There was a `}` missing.
I've fixed the code and made a PR on your branch on your repository.

I've also fixed the formatting.

If you could:

- [ ] double-check that function `invoice_replace_date_tags`
- [ ] see if my PR solves that problem

Then you can merge the PR and then tag me again. I'll make sure it ends up in InvoicePlane

**Review by nielsdrost7 — CHANGES_REQUESTED** [review]

**Comment by nielsdrost7** [comment]
@marneu I've created a PR for your repository on your branch https://github.com/marneu/InvoicePlane/pull/1
It should resolve some formatting issues and the end of function `invoice_replace_date_tags`

File: application/helpers/invoice_helper.php (hunk 1 above)
Comment (marneu) [reply application/helpers/invoice_helper.php:241]: Iv'e done fixes and these are productive in my 1.5 version now.
There are samples in the templates for usage and a debugShowVars.php to help for own templates.
A README within templates explains usage of the helpers.

Comment (marneu) [reply application/helpers/invoice_helper.php:240]: Do you see a chance for short communication on phone? (might be native tongue)

Comment (nielsdrost7) [reply application/helpers/invoice_helper.php:240]: Not today, maybe this weekend

File: application/helpers/invoice_helper.php [hunk 2]
```diff
@@ -128,4 +129,112 @@ function invoice_qrcode($invoice_id) {
...
+                $withReplacement = datefmt_format($dateFormat, $printDate);
+            }
+
+            // replace within item description
+            $item_description = str_replace($replaceThis, $withReplacement, $item_description);
+
+        } catch (\Exception $e) {
+            $item_description = str_replace($replaceThis, trans('invoice_replace_date_tag_invalid'), $item_description);
+        }
+
```
Comment (nielsdrost7) [inline application/helpers/invoice_helper.php:238]: Note to self:
This is where the `foreach()` ends

**Comment by nielsdrost7** [comment]
Duplicate with https://github.com/InvoicePlane/InvoicePlane/pull/1143

I'll close this one

**Pull request closed without merging**